pkg/kanx/kanx_test.go

```diff
@@ -0,0 +1,273 @@
+	c.Assert(err, IsNil)
+	c.Assert(p.GetPid(), Not(Equals), 0)
+	c.Assert(p.GetState(), Equals, ProcessState_PROCESS_STATE_RUNNING)
+	c.Assert(p.GetExitErr(), Equals, "")
+	c.Assert(p.GetExitCode(), Equals, int64(0))
+
+	ps, err := ListProcesses(ctx, addr)
+	c.Assert(err, IsNil)
+	c.Assert(ps, HasLen, 1)
+	c.Assert(ps[0].GetPid(), Equals, p.GetPid())
+	c.Assert(ps[0].GetState(), Equals, ProcessState_PROCESS_STATE_RUNNING)
+	c.Assert(ps[0].GetExitErr(), Equals, "")
+	c.Assert(ps[0].GetExitCode(), Equals, int64(0))
+
+	ctx = context.Background()
+	buf := bytes.NewBuffer(make([]byte, 0, 1024*1024))
+	isCancelled := false
+	go func() {
+		err := Stdout(ctx, addr, p.GetPid(), buf)
+		c.Assert(err, IsNil)
+		c.Assert(buf.String(), Equals, "")
+		c.Assert(isCancelled, Equals, true)
+	}()
+	sp, ok := server.pss.processes[p.GetPid()]
+	c.Assert(ok, Equals, true)
+	isCancelled = true
+	err = sp.cmd.Process.Kill()
+	c.Assert(err, IsNil)
+
+	buf.Reset()
+	err = Stdout(ctx, addr, p.GetPid(), buf)
+	c.Assert(err, IsNil)
+	c.Assert(buf.String(), Equals, "")
+
+	buf.Reset()
+	err = Stderr(ctx, addr, p.GetPid(), buf)
+	c.Assert(err, IsNil)
+	c.Assert(buf.String(), Equals, "")
+}
+
+func (s *KanXSuite) TestError(c *C) {
+	d := tmpDir(c)
+	addr := path.Join(d, "kanx.sock")
+	ctx, can := context.WithCancel(context.Background())
+	defer can()
+	server := newTestServer(d)
+	go func() {
+		err := server.Serve(ctx, addr)
+		c.Assert(err, IsNil)
+	}()
+	serverReady(ctx, addr, c)
+
+	p, err := CreateProcess(ctx, addr, "tail", []string{"-f", "/dev/null"})
+	c.Assert(err, IsNil)
+	c.Assert(p.GetPid(), Not(Equals), 0)
+	c.Assert(p.GetState(), Equals, ProcessState_PROCESS_STATE_RUNNING)
+	c.Assert(p.GetExitErr(), Equals, "")
+	c.Assert(p.GetExitCode(), Equals, int64(0))
+
+	ps, err := ListProcesses(ctx, addr)
+	c.Assert(err, IsNil)
+	c.Assert(ps, HasLen, 1)
+	c.Assert(ps[0].GetPid(), Equals, p.GetPid())
+	c.Assert(ps[0].GetState(), Equals, ProcessState_PROCESS_STATE_RUNNING)
+	c.Assert(ps[0].GetExitErr(), Equals, "")
+	c.Assert(ps[0].GetExitCode(), Equals, int64(0))
+
+	sp, ok := server.pss.processes[p.GetPid()]
+	c.Assert(ok, Equals, true)
+	err = sp.cmd.Process.Kill()
+	c.Assert(err, IsNil)
+
+	ps, err = ListProcesses(ctx, addr)
+	c.Assert(err, IsNil)
+	c.Assert(ps, HasLen, 1)
+	c.Assert(ps[0].GetPid(), Equals, p.GetPid())
+	c.Assert(ps[0].GetState(), Equals, ProcessState_PROCESS_STATE_FAILED)
+	c.Assert(ps[0].GetExitErr(), Equals, "signal: killed")
+	c.Assert(ps[0].GetExitCode(), Equals, int64(-1))
+
+	buf := bytes.NewBuffer(make([]byte, 0, 1024*1024))
+	err = Stdout(ctx, addr, p.GetPid(), buf)
+	c.Assert(err, IsNil)
+	c.Assert(buf.String(), Equals, "")
+
+	buf.Reset()
+	err = Stderr(ctx, addr, p.GetPid(), buf)
+	c.Assert(err, IsNil)
+	c.Assert(buf.String(), Equals, "")
+}
+
+var _ io.Writer = (*nilWriter)(nil)
+
+type nilWriter struct{}
+
+func (nw *nilWriter) Write(p []byte) (n int, err error) { return }
```

Comment:
Why not to use io.Discard instead?

Reply:
good suggestion. i hadn't seen io.Discard before

Reply:
@denisvmedia `io.Discard` is used in the new changes